Custom kernel for banded matmul in DSBCOO format

czox opened the pull request:
Currently, the banded format is not a full replacement of any of the existing distributed formats, it does not implement DSBSparse interface. Instead, it is an ad-hoc replacement for DSBCOO's `__matmul__` :
```
    def __matmul__(self, other: "DSBSparse") -> None:
        ...
        if self.allow_band_matmul and cuda_avail:
            return self.__band_matmul__(other)
        else:
            # proceed with default matmul
```
The `__band_matmul__` implements the on-the-spot conversion from DSBCOO to band, perform matmul in band format, and returns the result.

There is plenty of space for performance improvements - especially to minimize the number of back-and-forth format conversions, as well as the performance of the conversions themselves. Currently, this serves as a correctness testbed.

Commit:
banded matmul variant in COO. Work in progress

Commit:
band work in progress

Commit:
update

Commit:
cleanup

Commit:
fixed condidtional importing of torch and triton

Commit:
fixed conftest and imports in test_dbsparse

Commit:
fixed another torch import bug

alexnick83 commented on src/qttools/kernels/cuda/banded.py:
```diff
@@ -0,0 +1,1071 @@
+                b_ptrs += BLK_K * stride_bk
+            c = accumulator.to(tl.float16)
+        else:
+            for k in range(tl.cdiv(num_blocks * BLK_M, BLK_K)):
```
What is the actual difference with the first branch? I see two differences:
- There is no conversion to the destination datatype. If source and destination datatypes match, is there any actual overhead? If yes, then shouldn't all such variants use the second branch?
- There is no option to `allow_tf32`. I understand this flag can also affect float16 tensor cores (faster and less precise).

Btw, is there an option to target bfloat16?

alexnick83 replied:
Minor correction: I just saw that `allow_tf32` is not an exact match between PyTorch and Triton. Furthermore, Triton is deprecating `allow_tf32` in favor of `input_precision` (see https://triton-lang.org/main/python-api/generated/triton.language.dot.html). Please leave a TODO to update the call accordingly.

alexnick83 commented on src/qttools/kernels/cuda/banded.py:
```diff
@@ -0,0 +1,1071 @@
+        # and accumulate
+        # `a_ptrs` is a block of [BLOCK_SIZE_M, BLOCK_SIZE_K] pointers
+        # `b_ptrs` is a block of [BLOCK_SIZE_K, BLOCK_SIZE_N] pointers
+        # See above `Pointer Arithmetic` section for details
```
Is this your comment or from Triton's tutorial?

alexnick83 commented on src/qttools/kernels/cuda/banded.py:
```diff
@@ -0,0 +1,1071 @@
+                b_row = tl.load(b_row_ptrs)  # [BLOCK_SIZE_N]
+
+                # Outer product accumulation
+                accumulator += tl.where(
```
Aren't `offs_am` and `offs_bn` by definition (lines 905-906) less than M? Is it possible that this is killing performance?

alexnick83 commented on tests/datastructures/test_dsbsparse.py:
```diff
@@ -1,25 +1,65 @@
 from qttools.utils.mpi_utils import get_section_sizes
 
+cuda_avail = False
+if "torch" in sys.modules:
```
This is probably fine(?), but the canonical way is:
```python
try:
    import torch
    ...
except (ImportError, ModuleNotFound):
    pass

alexnick83 commented on tests/datastructures/test_dsbsparse.py:
```diff
@@ -1,25 +1,65 @@
-def _create_coo(sizes: NDArray) -> sparse.coo_matrix:
+
+def _create_coo(sizes: NDArray, dtype=xp.complex128) -> sparse.coo_matrix:
     """Returns a random complex sparse array."""
```
Docstring needs update (does not necessarily return a sparse matrix any more).

alexnick83 commented on tests/datastructures/test_dsbsparse.py:
```diff
@@ -173,7 +213,7 @@ def test_getitem(
         dense = dsbsparse.to_dense()
 
-        reference = dense[..., *accessed_element]
+        reference = dense[(..., *accessed_element)]
```
Why are all the indexed accesses below converted to tuples? It should be semantically the same, but was there some version incompatibility?

alexnick83 reviewed (changes requested):
Looks good overall. I have a few questions and minor suggestions, mostly to add extra comments.

vincent-maillou reviewed (commented):
If I understand correctly this implementation will serve as testbed to compare against a proper `dsbbanded` implementation of the data structure? Should this be merged on `dev`?
I would expect `dsbbanded` to be suited for merge maybe not the test-bed.

vetschn reviewed (changes requested):
So i can't say too much about the actual triton kernel implementation but i can give some more general comments since my review was requested.

Before we can merge this into `dev` the whole banded matmul needs to be refactored out from the datastructure directly. It needs to be a routine in a separate module (something like `src/qttools/datastructures/routines.py`) with a signature like
```python
def banded_matmul(a: DSBSparse, b: DSBSparse, out: DSBSParse, **kwargs) -> None:
    ...
```
The `kwargs` can be all the required parameters like `BLK_*`, `allow_tf32`, etc. Ultimately the `dsbcoo.py` module should stay completely unmodified if not for some reason _absolutely_ necessary.

The current  `DSBSparse.__matmul__` method was only there for testing the correctness of all the physics surrounding this matmul part in the code, but this method should eventually be removed completely from the datastructure since it always allocates new memory (which is not feasible for bigger runs). Matrix multiplication should be handled through separate routines that write to a preallocated output buffer (that has the correct sparsity structure already).

Also like @vincent-maillou already mentioned we can't use the `.todense()` method in the context of an actual simulation, even for the tiniest structures this is likely to cause issues :grimacing:

Apologies for the confusion, we should have made this stuff clear earlier

alexnick83 committed:
Created new file `qttools/datastructures/routines.py`.

The routines file should host routines that operate among DSBSparse arrays and do not fit in the specifications of the datastructure classes. The first routine included is `banded_matmul` (@czox), which implemented banded matrix-multiplication using Triton.

alexnick83 committed:
Fixes for copying back from dense representation to sparse.

alexnick83 committed:
Refactoring DSBCOO class to use the methods in the routines flie.

alexnick83 committed:
Merge branch 'dev' into band_matmul_dev

alexnick83 committed:
Fixes for banded matmul with more than 1 batch dimensions.

alexnick83 committed:
Fix for inability to directly compare the None slice with a slice containing GPU scalars.

alexnick83 committed:
Removed uncessary stack access.

alexnick83 committed:
Clean-up.

alexnick83 committed:
Put output annotation in quote marks to avoid test error when torch is not installed.

alexnick83 committed:
Reworked banded matmul tests to use the related routine directly.

alexnick83 committed:
Clean-up

alexnick83 committed:
Restored torch check.
